tests/test262_runner.rs

```diff
@@ -0,0 +1,589 @@
+use serde::{Deserialize, Serialize};
+use std::{
+    borrow::Cow,
+    collections::HashMap,
+    ffi::OsStr,
+    fs::{read_dir, File},
+    io::{ErrorKind, Read},
+    path::PathBuf,
+    process::{Command, Stdio},
+    time::Duration,
+};
+use wait_timeout::ChildExt;
+
+#[derive(Deserialize, Serialize, Debug, Clone, Copy, PartialEq, Eq)]
+#[serde(rename_all = "UPPERCASE")]
+enum TestExpectation {
+    Pass,
+    Fail,
+    Crash,
+    Timeout,
+}
+
+/// These directory names are filtered out by default.
+const SKIP_DIRS: &[&str] = &[
+    "annexB", "intl402", "staging",
+    //
+    //"built-ins",
+    //"language",
+];
+
+fn is_test_file(file_name: &str) -> bool {
+    file_name.ends_with(".js") && !file_name.contains("_FIXTURE")
+}
+
+#[derive(Debug)]
+struct Test262Runner {
+    runner_base_path: PathBuf,
+    tests_base: PathBuf,
+    nova_harness_path: PathBuf,
+    nova_cli_path: PathBuf,
+    expectations: HashMap<PathBuf, TestExpectation>,
+    treat_crashes_as_failures: bool,
+
+    num_tests_run: usize,
+    unexpected_results: HashMap<PathBuf, TestExpectation>,
+}
+
+impl Test262Runner {
+    const TEST_TIMEOUT: Duration = Duration::from_secs(10);
+
+    /// This error code denotes a parse error or an uncaught exception, all
+    /// others are treated as crashes.
+    const FAILURE_ERROR_CODE: i32 = 1;
+
+    /// If `filters` is empty, it means run every test. Otherwise, only tests
+    /// which have one of the entries of `filters` as its prefix should run.
+    pub fn run(&mut self, filters: &[PathBuf]) {
+        self.walk_dir(&self.tests_base.clone(), filters);
+    }
+
+    /// If `filters` is empty, every test in this directory and its descendants
+    /// should be run. Otherwise, only tests which have one of the entries of
+    /// `filters` as its prefix (considering their relative path to the `path`
+    /// directory) will be run.
+    fn walk_dir(&mut self, path: &PathBuf, filters: &[PathBuf]) {
+        for entry in read_dir(path).unwrap() {
+            let entry = entry.unwrap();
+
+            if entry.file_type().unwrap().is_dir() {
+                if let Some(child_filters) = Self::filter_dir(filters, &entry.file_name()) {
+                    self.walk_dir(&entry.path(), &child_filters);
+                }
+            }
+
+            if entry.file_type().unwrap().is_file()
+                && Self::filter_file(filters, &entry.file_name())
+            {
+                self.run_test(&entry.path());
+            }
+        }
+    }
+
+    fn filter_dir(filters: &[PathBuf], os_file_name: &OsStr) -> Option<Box<[PathBuf]>> {
+        if SKIP_DIRS.contains(&os_file_name.to_str().unwrap()) {
+            return None;
+        }
+        if filters.is_empty() {
+            return Some(vec![].into_boxed_slice());
+        }
+
+        let mut child_filters = Vec::with_capacity(filters.len());
+        for filter in filters {
+            if filter == OsStr::new("") || filter == os_file_name {
```

Comment:
**question:** What is the purpose of this if check case and the early return? Is this some error condition? Maybe add a small comment here.